### PR TITLE
PLT-322 Updated direct channels

### PR DESCRIPTION
**Pull request by hmhealey**
Does a bunch of things:
1) Adds Preferences table which is used to store each users' preferences, separate from the user object. Each preference has a category, name, an optional alternate id, and a value. The alt id can be used for things like assigning a preference to a specific channel or user.
2) Changes the Direct Messages list on the left sidebar to be controlled by a preference instead of populated based on who you've received messages from.
3) Adds the ability to add/remove people from your Direct Messages list

Note that there's still some styling to be done by Asaad and some additional UI changes to the More Direct Channels modal that were included in the spec that will be implemented likely after 1.1.0

### CODE REVIEW

**Comment by hmhealey** [comment]
Sorted out the build gremlins finally and the tests that were failing on Postgres. Ready for review.

**Comment by crspeller** [comment]
@hmhealey Needs rebase.

File: api/preference.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	decoder := json.NewDecoder(r.Body)
+	if err := decoder.Decode(&preferences); err != nil {
+		c.Err = model.NewAppError("setPreferences", "Unable to decode preferences from request", err.Error())
+		c.Err.StatusCode = http.StatusBadRequest
+		return
+	}
+
+	// just attempt to save/update them one by one and abort if one fails
+	// in the future, this could probably be done in a transaction, but that's unnecessary now
+	for _, preference := range preferences {
```
Comment (crspeller) [inline api/preference.go]: Failing half way though could cause some strange state.

File: api/preference.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+	}
+
+	w.Write([]byte("true"))
+}
+
+func getPreferencesByName(c *Context, w http.ResponseWriter, r *http.Request) {
+	params := mux.Vars(r)
+	category := params["category"]
+	name := params["name"]
+
```
Comment (crspeller) [inline api/preference.go]: Permissions check?

File: api/preference.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+	if result := <-Srv.Store.Preference().GetByName(c.Session.UserId, category, name); result.Err != nil {
+		c.Err = result.Err
+		return
+	} else {
+		data := result.Data.([]*model.Preference)
+
+		if len(data) == 0 {
+			if category == model.PREFERENCE_CATEGORY_DIRECT_CHANNELS && name == model.PREFERENCE_NAME_SHOW {
```
Comment (crspeller) [inline api/preference.go]: Why is this necessary? Why to preferences care about what they specify?

File: api/user.go [hunk 4]
```diff
@@ -218,6 +220,53 @@ func CreateUser(c *Context, team *model.Team, user *model.User) *model.User {
 	}
 }
 
+func fireAndForgetAddDirectChannels(user *model.User, team *model.Team) {
+	go func() {
+		AddDirectChannels(user.Id, team.Id)
+	}()
+}
+
+func AddDirectChannels(userId, teamId string) []*model.Preference {
```
Comment (crspeller) [inline api/user.go]: Why is this necessary? Why not just default to show if no preference?

File: model/preference.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+package model
+
+import (
+	"encoding/json"
+	"io"
+)
+
+const (
+	PREFERENCE_CATEGORY_DIRECT_CHANNELS = "direct_channels"
+	PREFERENCE_CATEGORY_TEST            = "test" // do not use, just for testing uniqueness while there's only one real category
```
Comment (crspeller) [inline model/preference.go]: Why is this necessary outside unit tests?

File: api/preference.go (hunk 2 above)
Comment (hmhealey) [inline api/preference.go]: Do we need that? It relies on having a Context with a Session and users are only ever to get their own preferences from that

File: api/preference.go (hunk 1 above)
Comment (hmhealey) [inline api/preference.go]: Oh yeah that's true. I'll move this into the store and wrap it in a transaction so it's all or nothing

**Comment by crspeller** [comment]
@hmhealey You have travis failures.

File: api/preference.go (hunk 3 above)
Comment (hmhealey) [inline api/preference.go]: This is just so that we can lazily add default channels for any users that existed before I added this so that their sidebar isn't empty. I did it this way based on the discussion we had yesterday in dev chat and a followup chat that I had with Corey

File: api/user.go (hunk 4 above)
Comment (hmhealey) [inline api/user.go]: We don't want to show all users in the sidebar by default or else the list will become massive for anything other than small teams. We can just pick 10 users semi-randomly to fill out the list so that it is populated and not empty. 

This is actually redundant if you consider the lazy version that we do when the getPreferences api is called so I might just remove doing it here.

File: model/preference.go (hunk 5 above)
Comment (hmhealey) [inline model/preference.go]: We need these for the IsXValid functions below. Those are called by the Store functions so there's no way of bypassing them.

**Comment by hmhealey** [comment]
Going to close this while I make those changes, fix some bugs I found, and probably squash it. Will resubmit once that's done.
